--- a/ppdc/Makefile
+++ b/ppdc/Makefile
@@ -182,13 +182,7 @@ local:	genstrings
 
 genstrings:		genstrings.o libcupsppdc.a ../cups/$(LIBCUPSSTATIC) \
 			sample.drv ../data/media.defs
-	echo Linking $@...
-	$(LD_CXX) $(ARCHFLAGS) $(ALL_LDFLAGS) -o genstrings genstrings.o \
-		libcupsppdc.a $(LINKCUPSSTATIC)
-	$(CODE_SIGN) -s "$(CODE_SIGN_IDENTITY)" $@
-	echo Generating localization strings...
-	./genstrings >sample.c
-
+	echo Not Linking $@...
 
 #
 # ppdc, the PPD compiler.
