From c0befab33abe1476625311a4f5495152cdb37b17 Mon Sep 17 00:00:00 2001
From: Florian Knodt <adlerweb@adlerweb.info>
Date: Fri, 28 Feb 2020 23:13:37 +0100
Subject: [PATCH 2/6] Add missing backend header; fixes cupsBackChannelRead
 error

---
 src/common/CupsPrintEnvironment.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/common/CupsPrintEnvironment.h b/src/common/CupsPrintEnvironment.h
index 7dadb0f..a61ef68 100644
--- a/src/common/CupsPrintEnvironment.h
+++ b/src/common/CupsPrintEnvironment.h
@@ -21,6 +21,8 @@
 #ifndef h952b1c81_8931_433a_8479_7ae6d8e85a86
 #define h952b1c81_8931_433a_8479_7ae6d8e85a86
 
+#include <cups/backend.h>
+#include <cups/sidechannel.h>
 #include "PrinterDriver.h"
 
 namespace DymoPrinterDriver
-- 
2.25.1

